### DISCUSSION TITLE
Sentence classification

### DISCUSSION THREAD

**Discussion by lmanhes** [explosion/spaCy#6836] [Help: Coding & Implementations] [answered]
Hi,

I would like to create a custom pipeline component to classify each Span (sentence) of a Doc (book-like document).

I already have a trained TextCategorizer model but it seems that it can only be used on a Doc and not on a Span. I also tried to use as_doc() to transform each sentence into a Doc but it's way too long. 

Is there a way to use this model on Span?

**Comment by svlandeg** [comment]
What do you mean `span.as_doc()` is "too long", does it significantly slow down the script?

**Comment by lmanhes** [comment, marked as answer]
I found a workaround, after adding the TextCategorizer component in the pipeline (my_textcategorizer), I created a custom component such as:

```python
class SpanClassifier(object):

    def __init__(self, nlp: Language):
        self.nlp = nlp

        Span.set_extension("sentiment", default=None, force=True)

        self.sentiment_clf = core_model.get_pipe("my_textcategorizer")

    def __call__(self, doc: Doc) -> Doc:
        sentences = list(doc.sents)

        for batch_sents in minibatch(sentences, size=128):
            batch_sents = list(batch_sents)

            sentiment_scores, _ = self.sentiment_clf.predict(batch_sents)
            sentiment_idx = np.argmax(sentiment_scores, -1)

            for i, sent in enumerate(batch_sents):
                sent._.sentiment = self.sentiment_clf.labels[sentiment_idx[i]]

        return doc
```
